### PR TITLE
refactor build interfaces

### PR DESCRIPTION
**Pull request by bparees**
Attempting to bring some structure to the build interfaces we had scattered about the code...  the intent is to define all the interfaces and the "standard" implementation in the util.go file, define the mock impls in the test cases that use them, and reduce the number of interfaces (rather than having an interface per build operation, have an interface that aggregates all the build operations).

this makes it slightly harder to mock the interface since you have to provide more dummy implementations, but it makes it much easier (imho) to understand the set of interfaces involved in the build process and where they're implemented.

### CODE REVIEW

**Comment by bparees** [comment]
@soltysh @mfojtik @csrwng PTAL

i think there is more that can be done here, but again i'd like to get this iteration in to avoid rebase/merge conflict nightmares

**Comment by bparees** [comment]
[test]

File: pkg/build/builder/util.go [hunk 1]
```diff
@@ -26,3 +30,67 @@ func getBuildEnvVars(build *api.Build) map[string]string {
 	}
 	return envVars
 }
+
+type BuildConfigClientInterface interface {
```
Comment (csrwng) [inline pkg/build/builder/util.go]: godoc?

File: pkg/build/builder/util.go [hunk 2]
```diff
@@ -26,3 +30,67 @@ func getBuildEnvVars(build *api.Build) map[string]string {
...
+	CreateBuild(namespace string, build *buildapi.Build) (*buildapi.Build, error)
+	CreateBuildFromConfig(config *buildapi.BuildConfig, revision *buildapi.SourceRevision, imageSubstitutions map[string]string) error
+	CreateBuildWithImmutableId(config *buildapi.BuildConfig) error
+	UpdateBuild(namespace string, build *buildapi.Build) (*buildapi.Build, error)
+}
+
+// OSClientBuildClient is a BuildClientInterface which delegates to the OpenShift client interfaces
+type OSClientBuildClient struct {
+	Client osclient.Interface
+}
```
Comment (csrwng) [inline pkg/build/builder/util.go]: Maybe there should be a NewXXX function to create an instance of this client that returns a BuildClient interface

**Comment by csrwng** [comment]
The builder package seems like an odd place to put the interfaces, maybe they should go in their own client package?

File: pkg/build/builder/util.go [hunk 3]
```diff
@@ -26,3 +30,67 @@ func getBuildEnvVars(build *api.Build) map[string]string {
 	}
 	return envVars
 }
+
+type BuildConfigClientInterface interface {
+	GetBuildConfig(namespace, name string) (*buildapi.BuildConfig, error)
+	UpdateBuildConfig(buildConfig *buildapi.BuildConfig) error
+}
+
+type BuildClientInterface interface {
```
Comment (smarterclayton) [inline pkg/build/builder/util.go]: So we don't do this for a reason - to make local testability easier.  This is kind of going against the conventions we're using in both Kubernetes and Origin.

File: pkg/build/builder/util.go [hunk 4]
```diff
@@ -26,3 +30,67 @@ func getBuildEnvVars(build *api.Build) map[string]string {
...
+	CreateBuild(namespace string, build *buildapi.Build) (*buildapi.Build, error)
+	CreateBuildFromConfig(config *buildapi.BuildConfig, revision *buildapi.SourceRevision, imageSubstitutions map[string]string) error
+	CreateBuildWithImmutableId(config *buildapi.BuildConfig) error
+	UpdateBuild(namespace string, build *buildapi.Build) (*buildapi.Build, error)
+}
+
+// OSClientBuildClient is a BuildClientInterface which delegates to the OpenShift client interfaces
+type OSClientBuildClient struct {
```
Comment (smarterclayton) [inline pkg/build/builder/util.go]: This whole type is an anti pattern - creating a wrapper to use by arbitrary consumers in arbitrary places.

File: pkg/build/controller/controller_test.go [hunk 5]
```diff
@@ -8,22 +8,49 @@ import (
...
+type okBuildClient struct{}
 
-func (obu *okBuildUpdater) UpdateBuild(namespace string, build *buildapi.Build) (*buildapi.Build, error) {
+func (okc *okBuildClient) UpdateBuild(namespace string, build *buildapi.Build) (*buildapi.Build, error) {
 	return &buildapi.Build{}, nil
 }
 
-type errBuildUpdater struct{}
+// dummy methods to finish implementing the interface
+func (okc *okBuildClient) CreateBuild(namespace string, build *buildapi.Build) (*buildapi.Build, error) {
```
Comment (smarterclayton) [inline pkg/build/controller/controller_test.go]: This is the problem with this pattern - a bunch of empty methods that don't convey useful info to an implementor.  Which is why we don't do this...

**Comment by smarterclayton** [comment]
I'm not in favor of this change given our existing conventions.

File: test/integration/buildclient_test.go [hunk 6]
```diff
@@ -195,13 +196,14 @@ func NewTestBuildOpenshift(t *testing.T) *testBuildOpenshift {
 
 	openshift.whPrefix = osPrefix + "/buildConfigHooks/"
 	osMux.Handle(openshift.whPrefix, http.StripPrefix(openshift.whPrefix,
-		webhook.NewController(clientWebhookInterface{osClient}, map[string]webhook.Plugin{
+		webhook.NewController(builder.OSClientBuildConfigClient{osClient}, builder.OSClientBuildClient{osClient}, map[string]webhook.Plugin{
```
Comment (smarterclayton) [inline test/integration/buildclient_test.go]: It would be more effective to implement a "real" implementation of ClientWebhookInterface that wraps a client in webhook.

File: pkg/build/controller/factory/factory.go [hunk 7]
```diff
@@ -14,16 +14,17 @@ import (
...
 	osclient "github.com/openshift/origin/pkg/client"
 	imageapi "github.com/openshift/origin/pkg/image/api"
 )
 
 type BuildControllerFactory struct {
-	Client              *osclient.Client
-	KubeClient          *kclient.Client
+	OSClient            osclient.Interface
+	KubeClient          kclient.Interface
+	BuildClient         builder.BuildClientInterface
```
Comment (smarterclayton) [inline pkg/build/controller/factory/factory.go]: This is worse than what was there before - now you have two interfaces providing the same behavior.  As with web hook, if you're looking to have a "canonical" client -> specific interface mapping, define it in the package that consumes it, and use it as close as possible to the place that consumes the interface.

Comment (smarterclayton) [inline pkg/build/controller/factory/factory.go]: Something like https://github.com/openshift/origin/blob/master/pkg/deploy/generator/config_generator.go#L33 is a good example of a multi-source yet mockable interface - Client is a "easy" implementation.  Or similar https://github.com/openshift/origin/blob/master/pkg/deploy/controller/factory/factory.go#L305

Both accomplish what you're trying to do here, but they avoid the "everyone has to mock methods they don't need" problem.

File: pkg/build/builder/util.go [hunk 8]
```diff
@@ -1,7 +1,11 @@
 package builder
 
 import (
-	"github.com/openshift/origin/pkg/build/api"
+	"github.com/golang/glog"
+
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	buildutil "github.com/openshift/origin/pkg/build/util"
+	osclient "github.com/openshift/origin/pkg/client"
```
Comment (mfojtik) [inline pkg/build/builder/util.go]: why 'osclient' and not just 'client'?

File: pkg/build/controller/controller_test.go [hunk 9]
```diff
@@ -8,22 +8,49 @@ import (
 	kerrors "github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
 
 	buildapi "github.com/openshift/origin/pkg/build/api"
+	builder "github.com/openshift/origin/pkg/build/builder"
```
Comment (mfojtik) [inline pkg/build/controller/controller_test.go]: is this needed?

File: pkg/cmd/server/origin/master.go [hunk 10]
```diff
@@ -151,6 +151,9 @@ func (c *MasterConfig) BuildClients() {
 func (c *MasterConfig) KubeClient() *kclient.Client {
 	return c.kubeClient
 }
+func (c *MasterConfig) OSClient() *osclient.Client {
```
Comment (mfojtik) [inline pkg/cmd/server/origin/master.go]: missing godoc for this and methods around

File: pkg/build/builder/util.go (hunk 8 above)
Comment (bparees) [inline pkg/build/builder/util.go]: clarity.  in some places we have the k8s client too, and i don't want to have to wonder which this is when looking at the code later, and have to jump up to the imports to get clarity.

Comment (mfojtik) [inline pkg/build/builder/util.go]: +1 care to fix it on other places as well?

File: pkg/build/builder/util.go (hunk 3 above)
Comment (bparees) [inline pkg/build/builder/util.go]: having seen what that has wrought, i really disagree with that convention.  it has created a nightmare of interfaces that make it really hard to follow what is being implemented, not to mention redundant logic all over the place.

I also don't really see that this reduces local testability, as you can see in the PR i've retained all the existing tests.

File: pkg/build/builder/util.go (hunk 4 above)
Comment (bparees) [inline pkg/build/builder/util.go]: This whole type already existed, all I did was rename it to something sensible and move it to a better location.

And I would beg to differ with it being used by "arbitrary" consumers (it's used by basically the entire system, it's the default implementation) and that it's in an arbitrary place, though i'm willing to entertain better places to put it....  but given that it's primairly used by stuff in the build pkg, having it in the build utils seemed about as good as anything.

File: pkg/build/builder/util.go [hunk 11]
```diff
@@ -26,3 +30,67 @@ func getBuildEnvVars(build *api.Build) map[string]string {
...
+		return err
+	}
+	return nil
+}
+
+// TODO will implement this in a future pull, want to get the refactor delivered before
+// too much rebasing happens.
+// CreateBuildWithImmutableId creates a build using the OpenShift client and the current immutable imageid
+// from any ImageRepository that is associated to the BuildConfig by an ImageChangeTrigger
+func (c OSClientBuildClient) CreateBuildWithImmutableId(config *buildapi.BuildConfig) error {
```
Comment (soltysh) [inline pkg/build/builder/util.go]: I don't like that name, but unfortunately Go does not allow function overloading :disappointed:. But if we'll stick with it please change `Id` -> `ID` to be consistent with everywhere else.

File: pkg/build/builder/util.go (hunk 3 above)
Comment (soltysh) [inline pkg/build/builder/util.go]: :+1:

File: pkg/build/builder/util.go (hunk 2 above)
Comment (bparees) [inline pkg/build/builder/util.go]: I've seen that pattern elsewhere but i'm unclear on the benefit of:
a:=OSClientBuildClient{Client: myclient}
vs
a:=builder.NewOSClientBuildClient(myclient)
?
just so we get some parameter checking?  any other reason, assuming there's no actual logic beyond assignment in the constructor?

File: pkg/build/controller/controller_test.go (hunk 9 above)
Comment (bparees) [inline pkg/build/controller/controller_test.go]: no :)

File: pkg/build/builder/util.go (hunk 3 above)
Comment (soltysh) [inline pkg/build/builder/util.go]: Actually Ben's refactor matches upstream conventions, see https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/controller/replication_controller.go#L46-L51. The only problem I can see here, there's too much create build methods we're using.

Comment (soltysh) [inline pkg/build/builder/util.go]: This interface purpose is to provide external services to the controller. I'd propose moving `CreateBuildFromConfig` and `CreateBuildWithImmutableId` methods to the controller, because they contain logic that should be part of the controller. The only interaction with other parts of the system I see there is `CreateBuild` which is part of that interface and all the rest should happen inside controller IMHO.

File: pkg/build/controller/controller_test.go (hunk 5 above)
Comment (bparees) [inline pkg/build/controller/controller_test.go]: to me empty methods in test code is a lesser sin than the ugliness of a method per interface and the other baggage that comes with it.

File: pkg/build/builder/util.go (hunk 3 above)
Comment (soltysh) [inline pkg/build/builder/util.go]: Looking more into it, I haven't noticed this interface landed in `util.go`, sorry. I see a `BuildInterface` here which should contain create and update methods and the other two methods (`CreateBuildFromConfig`, `CreateBuildWithImmutableId`) are just plain util methods, that should not be part of that interface.

Comment (mfojtik) [inline pkg/build/builder/util.go]: @soltysh +1 i think the interfaces should be 'clean'

**Comment by soltysh** [comment]
I'm leaning towards @smarterclayton idea of not changing what we have here. What I'd propose is to come up with reasonable naming scheme for case such as [here](https://github.com/openshift/origin/blob/master/pkg/build/controller/controller.go#L34-L45). In there we have 3 different interfaces serving similar purpose, they differ only in the resource type they serve. I'd suggest renaming all of them to `<resource>Manager`, `<resource>Client` or whatever else as long as it'll be consistent in this file and among other controllers, my 0.02 :heavy_dollar_sign:

File: pkg/build/builder/util.go (hunk 8 above)
Comment (bparees) [inline pkg/build/builder/util.go]: that's a ton of places, i think i'd rather do that in a separate PR given the way this one is going.

File: pkg/build/builder/util.go (hunk 3 above)
Comment (bparees) [inline pkg/build/builder/util.go]: CreateBuildFromConfig and CreateBuildWithImmutableId need to be interfaces somewhere so they can be mocked...  I might be inclined to agree that the methods that "just" create the object w/ no other logic should be separated from the more sophisticated paths, though.

File: pkg/build/builder/util.go [hunk 12]
```diff
@@ -26,3 +30,67 @@ func getBuildEnvVars(build *api.Build) map[string]string {
...
+	return c.Client.Builds(namespace).Create(build)
+}
+
+// UpdateBuild updates build using the OpenShift client.
+func (c OSClientBuildClient) UpdateBuild(namespace string, build *buildapi.Build) (*buildapi.Build, error) {
+	return c.Client.Builds(namespace).Update(build)
+}
+
+// CreateBuild creates a build using the OpenShift client.
+func (c OSClientBuildClient) CreateBuildFromConfig(config *buildapi.BuildConfig, revision *buildapi.SourceRevision, imageSubstitutions map[string]string) error {
```
Comment (mfojtik) [inline pkg/build/builder/util.go]: godoc does not match

File: pkg/build/builder/util.go (hunk 1 above)
Comment (bparees) [inline pkg/build/builder/util.go]: added

File: pkg/build/builder/util.go (hunk 2 above)
Comment (bparees) [inline pkg/build/builder/util.go]: went ahead and did this, still not sure of the value though.

File: pkg/build/builder/util.go (hunk 11 above)
Comment (bparees) [inline pkg/build/builder/util.go]: changed

**Comment by bparees** [comment]
i've addressed the non controversial comments (will squash later), will see if i can find a compromise on the interface debate after some discussion with @mfojtik and @soltysh

File: pkg/build/builder/util.go (hunk 4 above)
Comment (bparees) [inline pkg/build/builder/util.go]: (to be clear i'm referring to the OSClientBuildClient type, now i'm not sure which type you were referring to.  if it was the BuildClientInterface, i guess we'll see if my round 2 updates after discussing w/ @mfojtik and @soltysh go over any better.

File: test/integration/buildclient_test.go (hunk 6 above)
Comment (bparees) [inline test/integration/buildclient_test.go]: There's no good reason for "ClientWebhookInterface" to exist, all it did was provide CreateBuild methods, there's nothing unique about that that demands the webhook component have its own version of the interface.  Which is why i've removed it and replaced it with the more generic BuildClient interface to be used by things that want to create/manipulate Builds and BuildConfigClient interface for things that want to create/manipulate BuildConfigs

File: pkg/build/controller/factory/factory.go (hunk 7 above)
Comment (bparees) [inline pkg/build/controller/factory/factory.go]: The logic behind this is that BuildControllerFactory creates BuildControllers, BuildControllers need to update Builds.  To provide abstraction for testing, we need to be able to create BuildControllerFactories that will create BuildControllers that use arbitrary mechanism to update Builds.  To accomplish that, I've provided a way to pass in an arbitrary BuildClient implementation.  

This allows for testing of the entire Factory implementation without having to mock an entire osclient.

As for locating the canonical implementation with the consumer, I can move the implementation into the factory, but master.go hardly seems like the right place (which is where there canonical implementation was located before)

**Comment by bparees** [comment]
@smarterclayton let me add my next commit reworking the interface definitions before we continue this discussion, i think it will alleviate at least some of your concerns.

File: pkg/build/controller/controller_test.go (hunk 5 above)
Comment (smarterclayton) [inline pkg/build/controller/controller_test.go]: I'm ok with well shared interfaces.  I'm not ok with abstracting those interfaces.  The code that uses an interface should define it - it's good to make those interfaces consistent across their usage, but creating a "spider" interface (in util! no less) that creates a web of unrelated connections for that unrelated function creates strictly more confusion.  Each consumer should take *exactly* the interface they use - not a generic interface.

**Comment by bparees** [comment]
The refactor has been refactored.  
Things I suspect @smarterclayton will still not like:
1) There is still a BuildClient and BuildConfigClient aggregate interface, though what they contain has been greatly stripped down.  However I did introduce individual interfaces for test mocking, so tests do not need to implement the full Build(Config)Client interface.

2) BuildControllerFactory still has an explicit BuildUpdater field rather than creating a BuildUpdater from the OSClient, as is being done for the "PodManager" and "ImageRepositoryClient".  I prefer my approach because (although not currently used for testing) it enabled one to create factories that will create BuildControllers that use the OSClient, but use an alternate implementation of BuildUpdater.  I don't really see any meaningful downside to having this be a part of the Factory (I would do it for PodManager and ImageRepositoryClient too, if it were up to me).  I also think "ControllerClient" is a pretty unclear concept/term, so that was another motivation to not follow that pattern here.

3) ImageChangeControllerFactory follows the same pattern as #2 

4) I moved the Build(Config)Client interfaces into a new package, "client".  I think you wanted that part, but you may not like that i left the canonical implementations in that package.  I think they make more sense there than anywhere else I might put them (eg putting them in "factory" which is the primary place they're used).  util.go is back to what it was.  it should probably go away completely, but that's beyond what i want to do here.  

5) We also still have a "util" package as well, which should probably go away, but it was not introduced by this refactor.

**Comment by bparees** [comment]
Removed the aggregate interfaces from point (1) above.

File: pkg/build/client/clients.go [hunk 13]
```diff
@@ -0,0 +1,69 @@
+package client
+
+import (
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	osclient "github.com/openshift/origin/pkg/client"
+)
+
+// BuildConfigGetInterface provides methods for getting BuildConfigs
```
Comment (smarterclayton) [inline pkg/build/client/clients.go]: Consider me mollified on the interfaces here.

It's probably not good to call these "Interface" - instead "Getter" or "Updater" is more appropriate.

File: pkg/build/controller/controller.go [hunk 14]
```diff
@@ -19,7 +20,7 @@ type BuildController struct {
 	BuildStore    cache.Store
 	NextBuild     func() *buildapi.Build
 	NextPod       func() *kapi.Pod
-	BuildUpdater  buildUpdater
+	BuildUpdater  buildclient.BuildUpdateInterface
```
Comment (smarterclayton) [inline pkg/build/controller/controller.go]: Reading this, `buildclient.BuildUpdater` looks better.

File: pkg/build/controller/factory/factory.go [hunk 15]
```diff
@@ -14,16 +14,17 @@ import (
...
 	strategy "github.com/openshift/origin/pkg/build/controller/strategy"
-	buildutil "github.com/openshift/origin/pkg/build/util"
 	osclient "github.com/openshift/origin/pkg/client"
 	imageapi "github.com/openshift/origin/pkg/image/api"
 )
 
 type BuildControllerFactory struct {
-	Client              *osclient.Client
-	KubeClient          *kclient.Client
+	OSClient            osclient.Interface
```
Comment (smarterclayton) [inline pkg/build/controller/factory/factory.go]: Good.

File: pkg/build/controller/factory/factory.go [hunk 16]
```diff
@@ -14,16 +14,17 @@ import (
...
+	KubeClient          kclient.Interface
+	BuildUpdater        buildclient.BuildUpdateInterface
```
Comment (smarterclayton) [inline pkg/build/controller/factory/factory.go]: Again, `buildclient.BuilderUpdater` would make me feel like you want me to be happy.

File: pkg/build/controller/image_change_controller.go [hunk 17]
```diff
@@ -54,6 +48,8 @@ func (c *ImageChangeController) HandleImageRepo() {
 				continue
 			}
 			icTrigger := trigger.ImageChange
+			// TODO this will currently match imagerepos with the same name but in a different
+			// namespace, we should be checking that the namespaces match.
 			if icTrigger.From.Name != imageRepo.Name {
```
Comment (smarterclayton) [inline pkg/build/controller/image_change_controller.go]: Why not fix the code?  

```
if icTrigger.From.Name != imageRepo.Name || (len(icTrigger.From.Namespace) != 0 && icTrigger.From.Namespace != imageRepo.Namespace) {
```

File: pkg/build/util/generate.go [hunk 18]
```diff
@@ -40,28 +45,36 @@ func GenerateBuildFromBuild(build *api.Build) *api.Build {
 	}
 }
 
+// TODO will implement this in a future pull, want to get the refactor delivered before
+// too much rebasing happens.
+// CreateBuildWithImmutableID creates a build using the OpenShift client and the current immutable imageid
+// from any ImageRepository that is associated to the BuildConfig by an ImageChangeTrigger
+func GenerateBuildWithImmutableID(config *buildapi.BuildConfig) *buildapi.Build {
```
Comment (smarterclayton) [inline pkg/build/util/generate.go]: If this is implemented in a future pull... why is it in this pull request?

File: pkg/build/webhook/controller.go [hunk 19]
```diff
@@ -42,8 +39,8 @@ type urlVars struct {
 }
 
 // NewController creates new webhook controller and feed it with provided plugins.
-func NewController(osClient webhookBuildInterface, plugins map[string]Plugin) http.Handler {
-	return &controller{osClient: osClient, plugins: plugins}
+func NewController(buildConfigGetter buildclient.BuildConfigGetInterface, buildCreator buildclient.BuildCreateInterface, plugins map[string]Plugin) http.Handler {
+	return &controller{buildConfigGetter: buildConfigGetter, buildCreator: buildCreator, plugins: plugins}
```
Comment (smarterclayton) [inline pkg/build/webhook/controller.go]: Break this into 3 lines

File: pkg/cmd/server/origin/master.go [hunk 20]
```diff
@@ -148,24 +148,42 @@ func (c *MasterConfig) BuildClients() {
 	c.osClient = osclient
 }
 
+// KubeClient returns the kubernetes client object
 func (c *MasterConfig) KubeClient() *kclient.Client {
 	return c.kubeClient
 }
+
+// OSClient returns the openshift client object
+func (c *MasterConfig) OSClient() *osclient.Client {
```
Comment (smarterclayton) [inline pkg/cmd/server/origin/master.go]: No, you can't do this.  You have to call a method that says what component is using the client.  There is no "generic use of the client", because eventually everyone stops using this and starts using the others.  Create a method for the name of the component that is acting.

**Comment by smarterclayton** [comment]
I like this version much better, a few name things and the client stuff left.

File: pkg/build/controller/controller.go (hunk 14 above)
Comment (bparees) [inline pkg/build/controller/controller.go]: not sure it does to me, but i'll change it.

File: pkg/build/client/clients.go (hunk 13 above)
Comment (bparees) [inline pkg/build/client/clients.go]: will do.

File: pkg/build/controller/factory/factory.go (hunk 16 above)
Comment (bparees) [inline pkg/build/controller/factory/factory.go]: i find it helpful when things are clearly spelled out as being interfaces w/o having to dig into them.  it tells me immediately that the implementation might be flexible and i need to consider why and how it changes, and tells me i can potentially change it if needed.  Having to jump to the definition to determine that is an extra step.

but again, i'll change it, i'm not that hung up on it, but i do think there are multiple valid schools of thought here.

File: pkg/build/util/generate.go (hunk 18 above)
Comment (bparees) [inline pkg/build/util/generate.go]: at the time it was easier to leave it in w/ the refactoring i'd done, now i'll pull it out.

File: pkg/build/controller/image_change_controller.go (hunk 17 above)
Comment (bparees) [inline pkg/build/controller/image_change_controller.go]: didn't stop to process your objectref changes and see that it could be fixed at this point, wasn't in scope for what i was trying to do, just wanted to make a note that it needed to be fixed so we didn't lose track of it.

will fix it.

File: pkg/cmd/server/origin/master.go (hunk 20 above)
Comment (bparees) [inline pkg/cmd/server/origin/master.go]: starts using what others?  i'll change it, but i think i need some education on the vision here.

**Comment by bparees** [comment]
Made the final changes in a separate commit, will squash if you're content.

File: pkg/cmd/server/origin/master.go [hunk 21]
```diff
@@ -481,14 +499,15 @@ func (c *MasterConfig) RunAssetServer() {
 // RunBuildController starts the build sync loop for builds and buildConfig processing.
 func (c *MasterConfig) RunBuildController() {
 	// initialize build controller
-	dockerImage := env("OPENSHIFT_DOCKER_BUILDER_IMAGE", "openshift/origin-docker-builder")
-	stiImage := env("OPENSHIFT_STI_BUILDER_IMAGE", "openshift/origin-sti-builder")
+	dockerImage := env("OPENSHIFT_DOCKER_buildclient._IMAGE", "openshift/origin-docker-builder")
+	stiImage := env("OPENSHIFT_STI_buildclient._IMAGE", "openshift/origin-sti-builder")
```
Comment (soltysh) [inline pkg/cmd/server/origin/master.go]: Was this change intentional?

File: pkg/build/client/clients.go [hunk 22]
```diff
@@ -0,0 +1,69 @@
+package client
```
Comment (soltysh) [inline pkg/build/client/clients.go]: I'd propose renaming the file to `client.go`, similar to what we have in `pkg/client/client.go`.

File: pkg/cmd/server/origin/master.go (hunk 21 above)
Comment (bparees) [inline pkg/cmd/server/origin/master.go]: no, global search/replace error.

**Comment by mfojtik** [comment]
LGTM :+1:

**Comment by soltysh** [comment]
One small nit, other than that LGTM :8ball:

File: pkg/build/client/clients.go (hunk 22 above)
Comment (bparees) [inline pkg/build/client/clients.go]: since there are multiple clients in this file, i think i prefer it as is.

**Comment by bparees** [comment]
---->squashed<----

**Comment by bparees** [comment]
@smarterclayton merge at will.

File: pkg/build/client/clients.go (hunk 22 above)
Comment (smarterclayton) [inline pkg/build/client/clients.go]: This is more of a convention thing.  If you have one file in a package, and you want to call the package a certain something, you generally call the only file in the package the same as the package.  That's usually the stronger convention.

File: pkg/cmd/server/origin/master.go [hunk 23]
```diff
@@ -148,24 +148,42 @@ func (c *MasterConfig) BuildClients() {
...
 	return c.kubeClient
 }
+
+// WebHookClient returns the webhook client object
 func (c *MasterConfig) WebHookClient() *osclient.Client {
 	return c.osClient
 }
+
+// BuildControllerClient returns the build controller client object
+func (c *MasterConfig) BuildControllerClient() *osclient.Client {
```
Comment (smarterclayton) [inline pkg/cmd/server/origin/master.go]: If you have BuildControllerClients, why do you need BuildControllerClient?

Comment (bparees) [inline pkg/cmd/server/origin/master.go]: because I don't want both and in the interest of your implied pending abstraction of these methods, I didn't think using the method that returned both and then assuming one of the two return values would always be the one i wanted, was the right thing to do.

File: pkg/build/client/clients.go (hunk 22 above)
Comment (bparees) [inline pkg/build/client/clients.go]: so as soon as we add a second file to this package, the name would be ok, right?

**Comment by bparees** [comment]
@smarterclayton removed the BuildControllerClient method.

**Comment by smarterclayton** [comment]
LGTM [merge]
